src/storage/garbage_collector.cpp

```diff
@@ -86,80 +120,331 @@ uint32_t GarbageCollector::ProcessUnlinkQueue() {
+      // curr is visible to txns traversed till now, so can't GC curr. next is not visible to any txn, attempt GC
+      // Since *active_txns_iter is not reading next, that means no one is reading this
+      // And so we can reclaim next
+      if (interval_length == 0) {
+        // This only happens when the first compaction is initiated. Other compactions begin in Case 3.
+        // This is the first undo record to compact
+        BeginCompaction(&start_record, curr, next, &interval_length);
+      } else {
+        // Process the undo record to determine the set of columns for the compacted record.
+        ReadUndoRecord(start_record, next, &interval_length);
+      }
+      // Update curr
+      curr = curr->Next();
+    } else {
+      if (interval_length > 1) {
+        // Create the undo record by a second traversal through the records to be compacted.
+        UndoRecord *compacted_undo_record = CreateUndoRecord(start_record, next);
+        // Copy the varlen attributes of the compacted record
+        CopyVarlen(compacted_undo_record);
+        // Link the compacted record to the version chain.
+        LinkCompactedUndoRecord(start_record, &curr, next, compacted_undo_record);
+      }
+      // Set interval length to 0
+      EndCompaction(&interval_length);
+      // Begin compaction for the next interval
+      BeginCompaction(&start_record, curr, next, &interval_length);
+      // curr was not claimed in the previous iteration, so possibly someone might use it
+      curr = curr->Next();
+    }
     next = curr->Next();
-    // This is a legitimate case where we truncated the version chain but had to restart because the previous head
-    // was aborted.
-    if (next == nullptr) return;
-    if (transaction::TransactionUtil::NewerThan(oldest, next->Timestamp().load())) break;
-    curr = next;
-  }
-  // The rest of the version chain must also be invisible to any running transactions since our version
-  // is newest-to-oldest sorted.
+  }
+
+  if (interval_length > 1) {
+    // Create the undo record by a second traversal through the records to be compacted.
+    UndoRecord *compacted_undo_record = CreateUndoRecord(start_record, next);
+    // Copy the varlen attributes of the compacted record
+    CopyVarlen(compacted_undo_record);
+    // Link the compacted record to the version chain.
+    LinkCompactedUndoRecord(start_record, &curr, next, compacted_undo_record);
+  } else if (interval_length == 1) {
+    if (active_txns_iter == active_txns->end()) {
+      // Exited out of previous loop such that the last record is not visible to any transaction
+      UnlinkUndoRecordVersion(start_record->Next());
+      start_record->Next().store(nullptr);
+    }
+  }
+  // Set interval length to 0
+  EndCompaction(&interval_length);
+
+  // active_trans_iter ends but there are still elements in the version chain. Can GC everything below
   curr->Next().store(nullptr);
```

Comment:
Sorry, I am not sure whether the function UnlinkUndoRecordVersion() will reclaim the variable UndoRecord itself. If the answer is YES, then the pointer curr here may be an empty pointer. If the answer is NO, where will it be reclaimed?